examples/riot/coap_channel/coap_channel.c

```diff
@@ -0,0 +1,284 @@
+  LF_DEBUG(NET, "CoapChannel: Try connect callback");
+  CoapChannel *self = _get_coap_channel_by_remote(remote);
+
+  if (memo->state == GCOAP_MEMO_TIMEOUT || coap_get_code_class(pdu) != COAP_CLASS_SUCCESS) {
+    self->state = NETWORK_CHANNEL_STATE_DISCONNECTED;
+    return;
+  }
+
+  // TODO: Analyse payload. The remote federate might respond with "false" it it is not open for connections or if it is
+  // already connected
+
+  self->state = NETWORK_CHANNEL_STATE_CONNECTED;
+}
+
+static lf_ret_t CoapChannel_try_connect(NetworkChannel *untyped_self) {
+  LF_DEBUG(NET, "CoapChannel: Try connect");
+  CoapChannel *self = (CoapChannel *)untyped_self;
+
+  /* Server */
+  // Do nothing. Every CoAP client is also a server in our case.
+  // If the client successfully connected to the server of the other federate,
+  // then we consider the connection as established
+  // The other federate will also connect to our server and after that consider
+  // the connection to us as established.
+
+  /* Client */
+  switch (self->state) {
+  case NETWORK_CHANNEL_STATE_CONNECTED:
+    return LF_OK;
+
+  case NETWORK_CHANNEL_STATE_OPEN:
+    if (!_send_coap_message(&self->remote, "/connect", _client_try_connect_callback)) {
+      LF_ERR(NET, "CoapChannel: try_connect: Failed to send CoAP message");
+      return LF_ERR;
+    }
+    return LF_IN_PROGRESS;
+
+  case NETWORK_CHANNEL_STATE_CONNECTION_IN_PROGRESS:
+    return LF_IN_PROGRESS;
+
+  case NETWORK_CHANNEL_STATE_DISCONNECTED:
+  case NETWORK_CHANNEL_STATE_LOST_CONNECTION:
+    self->state = NETWORK_CHANNEL_STATE_OPEN;
+    return LF_TRY_AGAIN;
+
+  case NETWORK_CHANNEL_STATE_UNINITIALIZED:
+    return LF_ERR;
+  }
+}
+
+static void CoapChannel_close_connection(NetworkChannel *untyped_self) {
+  LF_DEBUG(NET, "CoapChannel: Close connection");
+  CoapChannel *self = (CoapChannel *)untyped_self;
+
+  // TODO
+}
+
+static lf_ret_t CoapChannel_send_blocking(NetworkChannel *untyped_self, const FederateMessage *message) {
+  LF_DEBUG(NET, "CoapChannel: Send blocking");
+  CoapChannel *self = (CoapChannel *)untyped_self;
+
+  // TODO
+}
+
+static void CoapChannel_register_receive_callback(NetworkChannel *untyped_self,
+                                                  void (*receive_callback)(FederatedConnectionBundle *conn,
+                                                                           const FederateMessage *msg),
+                                                  FederatedConnectionBundle *conn) {
+  LF_INFO(NET, "CoapChannel: Register receive callback");
+  CoapChannel *self = (CoapChannel *)untyped_self;
+
+  self->receive_callback = receive_callback;
+  self->federated_connection = conn;
+}
+
+static void CoapChannel_free(NetworkChannel *untyped_self) {
+  LF_DEBUG(NET, "CoapChannel: Free");
+  CoapChannel *self = (CoapChannel *)untyped_self;
+
+  // TODO
+}
+
+void CoapChannel_ctor(CoapChannel *self, Environment *env, const char *remote_host) {
```

Comment:
So this is now hard-coded COAP over UDP? We might want to create a unique channel for each physical layer that all shares some common COAP implementation. `UdpCoapChannel`, `BleCoapChannel`, `ZigbeeCoapChannel`.

Reply:
For now yes, I need to see how to abstract the protocol (connection, endpoint logic) from the underlying layers when I explore them deeper as a next step.
RIOT is in a special place here as it likes to run IP over almost everything :D

Reply:
After our discussion, and some exploration I think that hardcoding it ontop of UDP makes sense since this seems to be provided for many networking protocols on both Zephyr and RIOT